### DISCUSSION TITLE
Plugin support

### DISCUSSION THREAD

**Discussion by codingjoe** [takluyver/flit#663] [Ideas]
Hi there 👋

I maintain a couple of packages, that need several more build steps. Therefore, I have been writing a couple of extensions for flit recently. Like support for [gettext](https://github.com/codingjoe/flit-gettext) or [Sass](https://github.com/codingjoe/flit-sass).

However, I noticed that there currently isn't a good way to combine multiple plugins/extensions.

I'd love to have some form of registry for hooks. It doesn't need to be fancy, a simple list you can append functions to would suffice.

I will probably write a package for that. However, I wanted to pitch the idea upstream 🫡

Please reach out to me if that's something, you'd like me to contribute.

Cheers!
Joe

**Comment by takluyver** [comment]
Hi Joe! Thanks for your interest and your extensions. :slightly_smiling_face: 

This may seem odd, but it's kind of by design that Flit doesn't have any extension points. My aim is to have something simple, with as few ways to go wrong as possible, and my experience is that plugins always run counter to that - as soon as there's code from multiple projects running together, there's scope for incompatibilities. You can mitigate it by being careful with public vs. private APIs, versioned dependencies, and plenty of testing, but that's adding complexity to deal with complexity, and it's never perfect.

But seemingly the lack of any plugin system hasn't stopped people from making plugins anyway, which is simultaneously maddening and flattering, and I don't know what to do about it. :laughing: 

Perhaps I'll ask you a question: why build on Flit? When [PEP 621](https://peps.python.org/pep-0621/) standardised most of the contents of `pyproject.toml`, I figured Flit would probably fade into the background next to more powerful tools like Poetry & Hatch whose authors aren't so miserly with features as I am. Besides richer base feature sets, both of those have documented plugin support ([Poetry](https://python-poetry.org/docs/master/plugins/), [Hatch](https://hatch.pypa.io/latest/plugins/about/), also [PDM](https://pdm.fming.dev/latest/dev/write/)). I figured that the only reason to keep developing Flit is that sometimes, simplicity itself is a feature - e.g. for the libraries at the base of the Python packaging toolchain, `flit_core` is a build system with no dependencies of its own.

I assume my rationale isn't why you're using Flit, if you're happily writing plugins for it. So what is the attraction of Flit over the various alternatives?

**Reply by codingjoe** [reply]
Hi @takluyver, thanks for the enlightenment. Most interesting. I wasn't really aware of flits design principals. I haven't really used poetry or hatch yet, but I will look into it. Again, I appreciate you taking your time for such a detailed answer. Cheers! Joe

**Reply by codingjoe** [reply]
I also gotta admit, I wasn't aware of Hatch. Since both are run by PyPA, maybe it's a good idea to add a "distinguish" section or line to the docs.
… and don't get me started on Poetry…
